Seeding DB with DREAM Challenges

rrchai opened the pull request:
@tschaffter 
I pushed the codes and updated json files for seeding challenges. The followings properties still need to be tested if they work fine with current codes and latest schema.

- `start/endDate:`  they have been changed to optional in the latest schema, since some challenge has no start/end dates
- `organizerIds`, `dataProviderIds`, `grantIds`: check if empty array, `[]` is accepted in the latest services, since they are required

tschaffter commented:
@rrchai Thanks! I fixed the service issue that you reported about empty array (see https://github.com/Sage-Bionetworks/rocc-service/pull/113). I'll now try to find an elegant way to fix the seeding using RxJS.

tschaffter commented:
### ECONNRESET issue

I now experience `ECONNRESET` errors randomly. Initially after submitting a random number of persons, now also affecting other objects like Organization.

```
[HPM] Rewriting path from "/api/organizations?organizationId=clinical-proteomic-tumor-analysis-consortium" to "/organizations?organizationId=clinical-proteomic-tumor-analysis-consortium"
[HPM] POST /api/organizations?organizationId=clinical-proteomic-tumor-analysis-consortium ~> http://localhost:8080/api/v1
[HPM] Rewriting path from "/api/organizations?organizationId=columbia-university" to "/organizations?organizationId=columbia-university"
[HPM] POST /api/organizations?organizationId=columbia-university ~> http://localhost:8080/api/v1
[HPM] Error occurred while trying to proxy request /organizations?organizationId=cincinnati-childrens-hospital-medical-center from localhost:4200 to http://localhost:8080/api/v1 (ECONNRESET) (https://nodejs.org/api/errors.html#errors_common_system_errors)
[HPM] Error occurred while trying to proxy request /organizations?organizationId=celgene from localhost:4200 to http://localhost:8080/api/v1 (ECONNRESET) (https://nodejs.org/api/errors.html#errors_common_system_errors)
[HPM] Error occurred while trying to proxy request /organizations?organizationId=clinical-proteomic-tumor-analysis-consortium from localhost:4200 to http://localhost:8080/api/v1 (ECONNRESET) (https://nodejs.org/api/errors.html#errors_common_system_errors)
[HPM] Error occurred while trying to proxy request /organizations?organizationId=center-for-research-computing from localhost:4200 to http://localhost:8080/api/v1 (ECONNRESET) (https://nodejs.org/api/errors.html#errors_common_system_errors)
[HPM] Error occurred while trying to proxy request /organizations?organizationId=columbia-university from localhost:4200 to http://localhost:8080/api/v1 (ECONNRESET) (https://nodejs.org/api/errors.html#errors_common_system_errors)
[HPM] Rewriting path from "/api/organizations?organizationId=consejo-superior-de-investigaciones-cientificas" to "/organizations?organizationId=consejo-superior-de-investigaciones-cientificas"
[HPM] POST /api/organizations?organizationId=consejo-superior-de-investigaciones-cientificas ~> http://localhost:8080/api/v1
[HPM] Rewriting path from "/api/organizations?organizationId=corevitas" to "/organizations?organizationId=corevitas"
[HPM] POST /api/organizations?organizationId=corevitas ~> http://localhost:8080/api/v1
[HPM] Error occurred while trying to proxy request /organizations?organizationId=consejo-superior-de-investigaciones-cientificas from localhost:4200 to http://localhost:8080/api/v1 (ECONNRESET) (https://nodejs.org/api/errors.html#errors_common_system_errors)
[HPM] Error occurred while trying to proxy request /organizations?organizationId=corevitas from localhost:4200 to http://localhost:8080/api/v1 (ECONNRESET) (https://nodejs.org/api/errors.html#errors_common_system_errors)
[HPM] Rewriting path from "/api/organizations?organizationId=defense-advanced-research-projects-agency" to "/organizations?organizationId=defense-advanced-research-projects-agency"
[HPM] POST /api/organizations?organizationId=defense-advanced-research-projects-agency ~> http://localhost:8080/api/v1
[HPM] Rewriting path from "/api/organizations?organizationId=corrona" to "/organizations?organizationId=corrona"
[HPM] POST /api/organizations?organizationId=corrona ~> http://localhost:8080/api/v1
[HPM] Rewriting path from "/api/organizations?organizationId=dana-farber-cancer-institute" to "/organizations?organizationId=dana-farber-cancer-institute"
```

tschaffter commented:
@rrchai Getting this error after your last commit:

```
POSThttp://localhost:4200/api/organizations?organizationId=applied-proteogenomics-organizational-learning-and-outcomes-network
```

tschaffter commented:
Temporarily renaming this organization ID;

- `applied-proteogenomics-organizational-learning-and-outcomes-network` => `org-a`
- `eunice-kennedy-shriver-national-institute-of-child-health-and-human-development` => `org-b`

tschaffter commented:
@rrchai You replaced `Challenge.organizationIds` by `Challenge.organizations`, which is incorrect. I'll fix this.

tschaffter commented:
Removed the following organizations from challenge.json because their ID is too long (> 60 char).

- `eunice-kennedy-shriver-national-institute-of-child-health-and-human-development`
- `applied-proteogenomics-organizational-learning-and-outcomes-network`

rrchai commented:
Sorry about introducing new errors. There were some typos when I did the refactoring.

I have added following data (use other names to avoid conflicting your work): 
1. `organizations-fix.json`: trim the long names of organizations
2. `challenges-fix.json`: fix the incorrect property names; (this file is still person objects for organizers)

With @tschaffter suggestions in #84, I also added two files using fake ids:
4. `challenges-add-peronIds`: replace `organizers` with fake `organizerIds` to match the challenge schema.
3. `persons.json`: array of persons object following the order of `organizersId` in `challenges.json`.

note: if challenges do not have any persons, it will be the empty array in `organizerIds/organizers` in `challenges.json` and will not be written in `persons.json`.

tschaffter commented:
@rrchai I don't recommend to rely on the order of the objects in files. Instead, can you please generate fake ids for persons objects? If possible, please generate fake ids that comply with the format defined in schemas (e.g. [PersonId.yaml](https://github.com/Sage-Bionetworks/rocc-schemas/blob/main/openapi/components/parameters/PersonId.yaml)).

tschaffter commented:
@rrchai Also remove the Person objects from the Challenge object, include the fake ids instead. Please use the default parameter names like `organizerIds` or `grantIds` instead of what I did previously for the grants (`tmpGrantId`). It will take care of managing the fake and real ids in the seeding script.

tschaffter commented:
@rrchai You can generate "fake" MongoDB [ObjectId](https://pymongo.readthedocs.io/en/stable/api/bson/objectid.html) with the instruction `bson.objectid.ObjectId()` from the Python Package PyMongo.

rrchai commented:
@tschaffter  I have updated files:
- add `organizerId` to persons-fakeIds.json
- add `grantId` to grants-fakeIds.json
- update `organizerIds` and `grantIds` in challenge-fakeIds.json. Fake Ids in challenge objects match with the ids in person/grants objects.

tschaffter commented:
@rrchai Can you remove the following files?

- challenges-fakeIds.json
- grants-fakeIds.json
- organizations-fix.json
- persons-fakeIds.json

And update the remaining files so they meet the required defined in my previous comment. Ultimately we should only have the following files whose content match the ROCC schemas either based on the schemas `{schema}` (e.g. Tag, Organization) or `{schema}CreateRequest`.

- challenges.json
- grants.json
- organizations.json
- persons.json
- tags.json

tschaffter commented:
@rrchai @vpchung FYI Here is a [nice illustration](https://github.com/Sage-Bionetworks/rocc-app/pull/82/commits/6cf31a249589591bda3b4d61dd0a37c5666b8f99) on how Lodash can be used to manipulate objects. In this example, I use Lodash [omit method](https://lodash.com/docs/4.17.15#omit) to remove the property `id` from object in order to obtain `{schema}CreateRequest` objects that are ready to be posted. :)

tschaffter commented:
@rrchai @vpchung This PR is ready for your review. The remaining issue "Error: bundle initial-es5 exceeded maximum budget." should be fixed soon.

@rrchai I can't assign you as a reviewer because you create this PR. Do you mind giving a 👍 to this comment if you approve this PR?

vpchung commented on src/app/components/database-seed/database-seed.component.ts:
```diff
@@ -1,66 +1,214 @@
-import { mergeMap, tap } from 'rxjs/operators';
+import { forkJoin, Observable, of } from 'rxjs';
+import { map, mapTo, mergeMap, switchMap, tap } from 'rxjs/operators';
+import { merge as _merge, omit as _omit } from 'lodash';
```
Do we plan on updating Lodash as newer versions come out? According to the creator, [`omit` will be deprecated in v5 due to poor performance](https://github.com/lodash/lodash/issues/2930#issuecomment-272298477).  We may need to rewrite the code if so.

tschaffter replied:
We will update libraries every month to please the Dependabot God, unless the update of a given library is not trivial. I learned about the existence of `@babel/plugin-proposal-object-rest-spread` by following the link you shared. I don't find this approach suitable here as it depends on the order of the object properties. I decided to go with `_.pick` and specify the properties to keep (array). I wasn't sure about how `_.pick` handled array that include property names that are not in a given object, for example `Challenge.endDate` is optional. `_.pick` behave nicely and acts has a filter and it will not complain if a given object instance does not have the property. I'm still - and always - open to alternatives that you would like to share.

vpchung replied:
I like the `_.pick` approach as well! That being said, it may become quite tedious if numerous properties are needed... do you think that could happen?  If so, we could always write a util function that mimics `_.omit`?  I came across this article [here](https://levelup.gitconnected.com/omit-is-being-removed-in-lodash-5-c1db1de61eaf) that showcases different alternatives and their performances.

tschaffter replied:
Especially since we still are in the early stage of development, it's likely that object properties will be updated. It's also more likely that we will add or rename properties - i.e. which will need us to update the seeding code - than removing properties from `{schema}` object to get `{schema}CreateResponse`. Also one thing that I don't like about `_.pick` is its behavior as I described earlier can lead to a point where the list of properties specified is out of sync with the properties of `{schema}CreateResponse`.

> We can throw away the pickBy option as it it too slow and doesn’t provide any benefits.

Assuming that `pick` has similar performance, then let's use one of the alternative proposed. I just updated the code to an an `omit` function based on ESRest + JavaScript delete. This implementation does not allow to remove flatten paths, which should be fine for now.

vpchung reviewed (approved):
Looks good!  Just one design question about the future usage of Lodash.

rrchai reviewed (commented):
Looks great and I tested it works! @tschaffter
